src/client/Client.cc

```diff
@@ -10810,10 +10809,15 @@ void Client::C_Read_Sync_NonBlocking::finish(int r)
     if ((uint64_t)pos >= in->size)
       goto success;
 
+    clnt->put_cap_ref(in, CEPH_CAP_FILE_RD);
     {
       int have_caps2 = 0;
       r = clnt->get_caps(f, CEPH_CAP_FILE_RD, have_caps, &have_caps2, -1);
       if (r < 0) {
+	// Adjust cap_ref - do get_cap_ref again if get_caps fails. Otherwise,
+	// the cap_ref would go negative when C_Read_Finisher::finish_io does
+	// the final put_cap_ref
+        clnt->get_cap_ref(in, CEPH_CAP_FILE_RD);
```

Comment:
I remember a few months back when fixing the `Fc` cap issue in async i/o, IIRC @gregsfortytwo mentioned that is better to not incr/decr the cap ref counts based on conditional checks and on top of that, from looking at the `Client::get_caps` code, the reason for ret val -lt 0 can be many. Are we quite sure to incr the read cap ref based on the conditional check here?

Reply:
Wasn't that check

```
if (ref > 0) {
  --ref;
}
```

Which isn't the case there, isn't it?

Reply:
This case looked similar to it since it checks if the r < 0 then we incr/decr

Reply:
oh, i read below conversation now. all good